### PR TITLE
Fix Painless Lambdas for Java 9

### PR DESCRIPTION
**Pull request by jdconrad**
JDK9 has made changes to LambdaMetaFactory that prevent Painless lambdas from being able to do proper adaptation of types from the interface method to the delegate method even using a bridge.  In particular boxed types are no longer widened appropriately, and Object can no longer be adapted to primitive types.  Though, the former is likely a bug based on the spec, whether the latter is allowed according to spec is somewhat ambiguous.  

This fix adds a LambdaBootstrap class that replaces LambdaMetaFactory in Painless using the same methodology that LambdaMetaFactory was based on (http://cr.openjdk.java.net/~briangoetz/lambda/lambda-translation.html).  This allows us to do whatever adaptations we believe are necessary without having to worry about future changes/bugs in LambdaMetaFactory.

This is related to #23473.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java]: leftover?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java]: leftover?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java]: leftover?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java]: `{@link LambdaBootstrap}`? The advantage of the `@link` is that IDEs will automatically fix the comment if you rename the class.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Nor serialization which the link goes into a lot of depth about.

Also, one of the key reasons that they took that strategy for java code is that the compiler doesn't know what runtime it is going to be run under. invokedynamic future proofs the generated code. We don't have that concern. So I expect we can make different choices there too.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Reading this from top to bottom, I'm curious why we need this instead of calling the ctor directly.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: It is probably worth leaving a comment that this is done in the child classloader so this is a per-script counter.

Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: I think.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Can you indent this one extra time so it doesn't look like method body?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Chatted with @jdconrad - we have a different reason for needing invokedynamic than java does - we don't know the target functional interface until runtime sometimes. If you invoke a method on a `def` object that takes a lambda then we must delay building the class. For simplicity, we always delay.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java]: Accidental refactor.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java]: Accidental refactor.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java]: Accidental refactor.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java]: Added.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: To invoke a constructor the class would have to be newed and duped onto the stack.  The caller doesn't know what the generated class is, so it calls a factorymethod that can create the class instead.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Added a note about serialization.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Looks like that's not the case as when I test with two independent scripts the counter is incremented both times rather than starting at 1 for both of them.

**Comment by jdconrad** [comment]
@nik9000 Ready for another round -- think I responded to all comments.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java]: I'd either statically import all of these opcodes or none of them.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PSubCallInvoke.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PSubCallInvoke.java]: Leftover.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: s/List1/List/

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Yeah, there is only a single LambdaBootstrap so there is only one counter. I'd prefer one per script so I could look at the counter and know how many lambdas we've generated but I don't think that is worth holding up the PR for.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Can you move this method up with the other public method?

**Review by nik9000 — COMMENTED** [review]
I can't say I really understand all the subtleties but I think this makes sense.

I think it'd be worth testing what happens when you blow PIC and this falls back to megamorphic. I'm not sure if this code is worse or better than LambdaMetaFactory in that case but I'd love to understand it.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: It's not worth complicating this to have to reload this class on a per-script basis.  If you really want know the lambda count during debugging you can run one script at a time in a debugger.  Plus, you also already know the answer since you generate exactly one lambda factory method per lambda.

**Comment by jdconrad** [comment]
I don't understand this -- "I think it'd be worth testing what happens when you blow PIC and this falls back to megamorphic. I'm not sure if this code is worse or better than LambdaMetaFactory in that case but I'd love to understand it."  Can you please explain what you mean here?  Lambdas don't use this PIC at all except indirectly through another def type.  For that def type to even fall back to megamorphic requires the target type to change I believe at least 6 times.  (I'm not even sure this is possible using lambdas given our current supported whitelist.)  I also know the fallback is tested for other types already.  Maybe it's interesting in an academic sense, but it's really not practically necessary at this time.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Right, it isn't worth holding the PR for. I'd like it for debugging issues in running systems rather than the debugger. But it isn't a big deal.

**Comment by nik9000** [comment]
> Maybe it's interesting in an academic sense, but it's really not practically necessary at this time.

Sure. I'm debugging some issues someone pinged me about where our stack traces aren't good when we fall back to the megamorphic case so I'm probably overthinking this.

**Review by nik9000 — APPROVED** [review]
Let's get this in so we aren't busted in Java 9.

**Comment by jdconrad** [comment]
@nik9000 Is there an issue up for the stack traces?  I'm curious to see what the difference in stack traces looks like.

Also, thanks for the review!  I will make sure to respond to the rest of the comments before merging.

**Comment by nik9000** [comment]
> Is there an issue up for the stack traces? I'm curious to see what the difference in stack traces looks like.

I'm playing with it now. Someone pinged my privately. I'm still trying to reproduce.

**Comment by nik9000** [comment]
So yes, there is an issue with stack traces but it isn't unexpected. Basically the PIC and MIC end up in stacktraces when `def` is involved and that makes things hard to read. But you should be using `script_stack` to debug issues like these anyway.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Fixed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java]: Yes!  Fixed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PSubCallInvoke.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PSubCallInvoke.java]: Fixed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java]: Going to leave this one as the methods are in order of which ones will be called first.

**Comment by jdconrad** [comment]
@nik9000 Thank you for the review.

**Comment by nik9000** [comment]
Thanks for diving into LambdaMetaFactory!

**Comment by uschindler** [comment]
I figured out that the special handling of H_NEWINVOKESPECIAL is broken. The following test fails:

```java
    public void testCtorWithParams() {
        assertArrayEquals(new Object[] { "foo", "bar" },
                (Object[]) exec("List l = new ArrayList(); l.add('foo'); l.add('bar'); " +
                        "Stream stream = l.stream().map(StringBuilder::new);" +
                        "return stream.map(Object::toString).toArray()"));
    }
```

This code fails because the script returns a `Object[]` consisting only of 2 empty Strings. The reason for this is: `StringBuilder::new` points to the `StringBuilder(CharSequence)` ctor. The code in LambdaBootstrap ignores all parameters and calls the default ctor of `StringBuilder`. The result is mapped back to a String afterwards and is of course empty.

I have a fix ready...

**Comment by uschindler** [comment]
In addition, the static factory as replacement for the constrcutor is not needed. If you use `lookup.findConstructor()` everything works as it should. I will post both fixes in a PR.

**Comment by uschindler** [comment]
I opened PR #24406 to fix both issues. All tests pass for me with Java 8 and Java 9.

**Comment by uschindler** [comment]
I have seen some discussions initiated by @nik9000 and @jdconrad about the useless stack frames we see in the stack trace on exceptions. Since Java 8u60, the whole Lambda class is hidden from all stack traces (so user only sees a direct call in the stack trace to the lambda/function reference, but no wrapper classes.

The trick is to mark methods that you want to hide (because they are completely synthetic and have no useful meaning to users reading the stack). This would also solve the problem with the ever-incrementing class name counter.

This is the change in Java 8u60: http://hg.openjdk.java.net/jdk8u/jdk8u-dev/jdk/rev/f7dd864a52ea

We can do the same (after #24406 is merged):
```java
iface.visitAnnotation("Ljava/lang/invoke/LambdaForm$Hidden;", true);
```

The problem is that this class is internal only (OpenJDK specific). So I'd only add the annotation to the interface method and the ctor proxy only if `Class.forName("java/lang/invoke/LambdaForm$Hidden")` works. If you agree, I will provide a PR soon after #24406 is merged.

**Comment by uschindler** [comment]
I tried to hide the stack frames like done above. Unfortunately it does not seem to work if the classes are not anonymous (via Unsafe).

During that I found some smaller things, like the doPrivileged around getting the ctor: This is not needed, as it is public in our case. I can provide a PR later. I'd also change LambdaConversionException like the original LambdaMetaFactory on problems instead of IllegalStateException (which is wrong).

**Comment by uschindler** [comment]
As hiding the stack frames did not work, I made the counter "per-script". The trick was to put an `AtomicInteger` into the `Compiler.Loader` class, which is accessible from `LambdaBootstrap`. As it is no longer global, an integer is fine. I also removed the doPrivileged. See https://github.com/elastic/elasticsearch/pull/24406/commits/b547fa995e762b9bc3327da26c90ab828adfdc52

I don't like the IllegalStateExceptions in LambdaBootstrap instead of LambdaConversionExceptions. We should also change that. OpenJDKs bootstrap throws LambdaConversionException on every problem while creating the lambda class.

**Comment by uschindler** [comment]
Finally I also changed the codesource parameter in defineClass to be consistent in my PR. The lambdas are now running in same unprivileged code source.

**Review by uschindler — COMMENTED** [review]
I solved the mentioned issues in my other PR #24406

**Pull request merged**
